### PR TITLE
Final removal of uses of Windows wide char printing

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
There are existing uses of `%S`, but they remain only for Windows scenarios.

Replace uses of `%ls` with `%S` for consistency reasons.

/cc @jkotas @elinor-fung @janvorli

### CODE REVIEW

**Comment by BruceForstall** [comment]
I see you touched ilasm/ildasm. You should run the ilasm round-trip pipeline on the test when ready:

https://dev.azure.com/dnceng-public/public/_build?definitionId=140

**Comment by BruceForstall** [comment]
Is there some context here (link to an issue?) about why, and the strategy for the changes? E.g., Windows wide chars (UTF-16?) need to be converted to UTF-8 before being printed with `%s`?

File: src/coreclr/jit/disasm.cpp [hunk 1]
```diff
@@ -1229,12 +1229,12 @@ size_t CbDisassembleWithBytes(DIS* pdis, DIS::ADDR addr, const BYTE* pb, size_t
         if (fFirst)
         {
             pdis->CchFormatInstr(wz, ArrLen(wz));
-            fprintf(pfile, "%-*ls %ls\n", cchBytesMax, pwzBytes, wz);
+            fprintf(pfile, "%-*ls %S\n", cchBytesMax, pwzBytes, wz);
```
Comment (BruceForstall) [inline src/coreclr/jit/disasm.cpp:1232]: Should the other `ls` also change?

Comment (BruceForstall) [reply src/coreclr/jit/disasm.cpp:1232]: Also, this file is only ever compiled/run on Windows (and is probably bit-rotted). Should it even be touched?

**Comment by AaronRobinsonMSFT** [comment]
> Is there some context here (link to an issue?) about why, and the strategy for the changes? E.g., Windows wide chars (UTF-16?) need to be converted to UTF-8 before being printed with `%s`?

The context is about only using standards compliant `printf` when code is to run on any platform. The goal is to remove the need for the `printf` version that emulates Windows formatting rules that is [carried with the PAL](https://github.com/dotnet/runtime/blob/main/src/coreclr/pal/src/cruntime/printf.cpp). If the code is only expected to run on Windows and the change is overly complicated, then it remains. The biggest culprit in that bucket is some Watson logic. Other than that, if `printf` is being used the expectation is it only uses supported compliant flags. Alternative string printing with `%ls` or `%S` won't work due to issues with `wchar_t` being 4 bytes on non-Windows.

File: src/coreclr/tools/superpmi/superpmi-shared/asmdumper.cpp [hunk 2]
```diff
@@ -82,8 +82,8 @@ void ASMDumper::DumpToFile(HANDLE hFile, MethodContext* mc, CompileResult* cr)
 
         WCHAR instrMnemonic[64]; // I never know how much to allocate...
         disasm->CchFormatInstr(instrMnemonic, 64);
-        buff_offset += sprintf_s(&buff[buff_offset], bufflen - buff_offset, "\r\n%p %S",
-                                 (void*)((size_t)orig_hotCodeBlock + offset), instrMnemonic);
+        buff_offset += sprintf_s(&buff[buff_offset], bufflen - buff_offset, "\r\n%p %s",
+                                 (void*)((size_t)orig_hotCodeBlock + offset), ConvertToUtf8(instrMnemonic).c_str());
```
Comment (BruceForstall) [inline src/coreclr/tools/superpmi/superpmi-shared/asmdumper.cpp:86]: Once again, this is mostly-(probably)-dead code.

If not, isn't it expensive? It *looks* like you're not leaking memory...

**Review by BruceForstall — COMMENTED** [review]
The jit/spmi changes look basically ok but there's a question about why bother, especially with some cases that are currently not built.

**Comment by BruceForstall** [comment]
Is there a longer-term goal here beyond `printf`? Like making including the PAL compatible with non-Windows wchar_t and STL?

File: src/coreclr/jit/disasm.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/disasm.cpp:1232]: Oops. Missed that one.

**Comment by AaronRobinsonMSFT** [comment]
> Is there a longer-term goal here beyond printf? Like making including the PAL compatible with non-Windows wchar_t and STL?

Making the PAL compatible with `wchar_t` is highly unlikely although possible in the long run. Compat with STL is unlikely due to exception free constraints. However, perhaps the tenor of your question is slightly broader - platform provided headers files? That is a goal. Aspects of the STL, such as `std::unique_ptr<T>`, that don't throw exceptions should be consumable.

File: src/coreclr/jit/disasm.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/disasm.cpp:1232]: > Also, this file is only ever compiled/run on Windows (and is probably bit-rotted). Should it even be touched?

I would prefer to either remove the file entirely or change the format flags as appropriate.

**Comment by AaronRobinsonMSFT** [comment]
CI is green. @elinor-fung or @janvorli Please take a look.

File: src/coreclr/ilasm/asmparse.h [hunk 3]
```diff
@@ -197,8 +187,7 @@ class MappedFileStream : public ReadStream {
         return (m_hFile == INVALID_HANDLE_VALUE) ? NULL : map_file();
     }
 
-    const WCHAR* fileNameW;     // FileName (for error reporting)
-	//char	fileNameANSI[MAX_FILENAME_LENGTH*4];
+	char	fileNameUtf8[MAX_FILENAME_LENGTH*4]; // FileName (for error reporting)
```
Comment (janvorli) [inline src/coreclr/ilasm/asmparse.h:190]: Seeing a fixed buffer for file path always makes me feel uneasy. It seems that we could allocate it dynamically here without much hassle. Would you mind doing it that way?

**Review by janvorli — APPROVED** [review]
LGTM modulo one nit, thank you!

File: src/coreclr/debug/createdump/crashinfo.cpp [hunk 4]
```diff
@@ -929,6 +929,24 @@ FormatString(const char* format, ...)
...
+{
+    if (str == nullptr)
+        return{};
+
+    int len = WideCharToMultiByte(CP_UTF8, 0, str, -1, nullptr, 0, nullptr, nullptr);
+    if (len == 0)
+        return{};
+
+    ArrayHolder<char> buffer = new char[len + 1];
+    WideCharToMultiByte(CP_UTF8, 0, str, -1, buffer, -1, nullptr, nullptr);
```
Comment (elinor-fung) [inline src/coreclr/debug/createdump/crashinfo.cpp:946]: Shouldn't the `-1` for the `cbMultiByte` parameter be `len + 1`?

File: src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp [hunk 5]
```diff
@@ -183,13 +181,17 @@ void formatOutput(struct IDebugDataSpaces* memCallBack, ___in FILE* file, __inou
                                     MethodDescData.Request(g_sos,(CLRDATA_ADDRESS)arg);
 
                                     static WCHAR wszNameBuffer[1024]; // should be large enough
-                                    if (g_sos->GetMethodDescName(arg, 1024, wszNameBuffer, NULL) != S_OK)
+                                    static char szNameBuffer[ARRAY_SIZE(wszNameBuffer) * 3];
```
Comment (elinor-fung) [inline src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp:184]: + 1 for the buffer size?

*3 + 1 is the conversion to max length for UTF-8, right?

File: src/coreclr/debug/createdump/crashinfo.cpp (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/debug/createdump/crashinfo.cpp:946]: Yes.

File: src/coreclr/ilasm/asmparse.h (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/ilasm/asmparse.h:190]: Sure.

File: src/coreclr/debug/createdump/crashinfo.cpp (hunk 4 above)
Comment (elinor-fung) [reply src/coreclr/debug/createdump/crashinfo.cpp:946]: Bunch of other places too.

File: src/coreclr/vm/peimagelayout.cpp [hunk 6]
```diff
@@ -521,7 +521,11 @@ LoadedImageLayout::LoadedImageLayout(PEImage* pOwner, HRESULT* loadFailure)
     }
 
     IfFailThrow(Init(m_Module, true));
-    LOG((LF_LOADER, LL_INFO1000, "PEImage: Opened HMODULE %S\n", (LPCWSTR)pOwner->GetPath()));
+
+#ifdef LOGGING
+    MAKE_UTF8PTR_FROMWIDE(pathUtf8, pOwner->GetPath());
```
Comment (elinor-fung) [inline src/coreclr/vm/peimagelayout.cpp:526]: The PEImage path is `SString`, so you could just do `GetUTF8`.

File: src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp (hunk 5 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/StressLogAnalyzer/StressLogDump.cpp:184]: Yep

**Review by elinor-fung — APPROVED** [review]
Hurray!

**Pull request merged**